Improved handling of mode changes in buffextras.lua

someperson opened the pull request:
I made some changes to buffextras.lua to try to improve the handling of mode changes.

Here's what HexChat normally displays for some activity in a channel:
![1 - original](https://cloud.githubusercontent.com/assets/376450/18222076/49243efe-7151-11e6-9b83-26e5d06d0fb8.png)

And here is what buffextras turns this into in ZNC 1.6.3:
![2 - buffextras](https://cloud.githubusercontent.com/assets/376450/18222082/5fb01918-7151-11e6-9490-62ca2746d055.png)

The original buffextras.lua doesn't quite handle the mode changes properly. The mode changes from the server are unreadable, and the changes from the user clearly have a different format from what is normally used:
![3 - plugin](https://cloud.githubusercontent.com/assets/376450/18222090/a38bf968-7151-11e6-95b9-34996d763f43.png)

After my changes, this is the result:
![4 - plugin-mod](https://cloud.githubusercontent.com/assets/376450/18222092/aa40e9da-7151-11e6-84f1-1f14c9859278.png)

It's not perfect. Obviously, HexChat normally does some parsing of the mode changes and even breaks them apart so that it can display a nice message for particular changes (e.g. +b or +l). This is not implemented here, but I think it's at least an improvement.

Thanks for considering this change!

Sam

TingPing commented:
Should probably check `hexchat.prefs['irc_raw_modes']` also, if enabled use previous behavior.

someperson commented:
Thanks, I didn't know that preference existed. I just changed it so that it will use "Raw Modes" or "Channel Mode Generic" based on how it is set. Also cleaned it up a bit to avoid duplicate code.

TingPing commented:
Your usage of 'Channel Mode Generic' also seems wrong. It takes 4 args: nick, mode sign (+/-), mode letter, channel. You seem to combine the last 3 into one.

TingPing commented:
Also if you want to match HexChat behavior exactly we will have to expand this quite a bit. It parses specific letters into specific events like *Channel Ban* and falls back to *Channel Mode Generic*: https://github.com/hexchat/hexchat/blob/master/src/common/modes.c#L438-L572

Also when there are multiple modes (I don't know how znc handles this?) there is a different event *Channel Modes*.

You can see why I took the lazy route of *Raw Modes* :P

someperson commented:
I've now implemented support for the specific events. The modes are now also therefore split apart and sent separately, like HexChat does normally. I added a hook for 005, as you suggested would be possible, in order to retrieve the CHANMODES parameter, which affects parsing of arguments.

In theory, this should now pretty much emulate HexChat's normal behavior, though I haven't done much testing with it yet.

TingPing commented:
@someperson BTW the chanmodes property landed and has been out in a release for a few weeks.

TingPing commented on HexChat/buffextras.lua:
```diff
@@ -1,10 +1,6 @@
--- SPDX-License-Identifier: MIT
+--- SPDX-License-Identifier: MIT
```
What is this change? :P

someperson replied:
Oops. Will fix.

TingPing commented on HexChat/buffextras.lua:
```diff
@@ -46,13 +33,125 @@ hexchat.hook_server_attrs('PRIVMSG', function (word, word_eol, attrs)
+			emit('Raw Modes', name, string.format('%s %s', channel, modes))
 		else
-			emit('Kick', nick, word[6], channel, word_eol[9])
+			nickmodes = hexchat.props['nickmodes']
```
Should be `local`, same with others?

someperson replied:
Yep. Will fix.

FichteFoll commented:
What about using the `RECV` command and have Hexchat do its normal handling instead?

TingPing commented:
Needs rebased.
